src/maxwell/maxwell_eps.c

```diff
@@ -473,247 +614,158 @@ void set_maxwell_dielectric(maxwell_data *md,
-			   (eps_inv_mean.m02 - eps_mean_inv.m02) * norm2);
-		    x1 += ((eps_inv_mean.m01 - eps_mean_inv.m01) * norm0 +
-			   (eps_inv_mean.m12 - eps_mean_inv.m12) * norm2);
-		    x2 += ((eps_inv_mean.m02 - eps_mean_inv.m02) * norm0 +
-			   (eps_inv_mean.m12 - eps_mean_inv.m12) * norm1);
-
-		    md->eps_inv[xyz_index].m01 = (0.5*(x0*norm1 + x1*norm0)
-						  + eps_mean_inv.m01);
-		    md->eps_inv[xyz_index].m02 = (0.5*(x0*norm2 + x2*norm0)
-						  + eps_mean_inv.m02);
-		    md->eps_inv[xyz_index].m12 = (0.5*(x1*norm2 + x2*norm1)
-						  + eps_mean_inv.m12);
+           tau.m01 *= mesh_prod_inv;
+           tau.m02 *= mesh_prod_inv;
+           tau.m12 *= mesh_prod_inv;
 #endif
-	       }
-	       md->eps_inv[xyz_index].m00 = x0*norm0 + eps_mean_inv.m00;
-	       md->eps_inv[xyz_index].m11 = x1*norm1 + eps_mean_inv.m11;
-	       md->eps_inv[xyz_index].m22 = x2*norm2 + eps_mean_inv.m22;
+           
+           /* --- compute τ⁻¹[mean(τ(ε))] (i.e. the Kottke-averaged permittivity) --- */
+           /* τ⁻¹(τ) is defined by [Kottke PRE, Eq. (23)]:
+                         ( -1/τ₁₁    -τ₁₂/τ₁₁        -τ₁₃/τ₁₁       )
+                τ⁻¹(τ) = ( -τ₂₁/τ₁₁  τ₂₂-τ₂₁τ₁₂/τ₁₁  τ₂₃-τ₂₁τ₁₃/τ₁₁ ) = meanᴷ[ε]
+                         ( -τ₃₁/τ₁₁  τ₃₂-τ₃₁τ₁₂/τ₁₁  τ₃₃-τ₃₁τ₁₃/τ₁₁ )
+           */
+           eps_mean.m00 = -1/tau.m00;
+           eps_mean.m11 = tau.m11 - ESCALAR_NORMSQR(tau.m01)/tau.m00;
+           eps_mean.m22 = tau.m22 - ESCALAR_NORMSQR(tau.m02)/tau.m00;
+
+           ASSIGN_ESCALAR(eps_mean.m01,
+                    -ESCALAR_RE(tau.m01)/tau.m00, -ESCALAR_IM(tau.m01)/tau.m00);
+           ASSIGN_ESCALAR(eps_mean.m02,
+                    -ESCALAR_RE(tau.m02)/tau.m00, -ESCALAR_IM(tau.m02)/tau.m00);
+           ASSIGN_ESCALAR(eps_mean.m12,
+                    ESCALAR_RE(tau.m12) - ESCALAR_MULT_CONJ_RE(tau.m02, tau.m01)/tau.m00,
+                    ESCALAR_IM(tau.m12) - ESCALAR_MULT_CONJ_IM(tau.m02, tau.m01)/tau.m00 );
+
+           /* --- rotate eps_mean (i.e. τ⁻¹) back to the cartesian coordinate system --- */
+#define SWAP(a,b) { double xxx = a; a = b; b = xxx; }  /* invert via tranposition */  
+           SWAP(Rot[0][1], Rot[1][0]);
+           SWAP(Rot[0][2], Rot[2][0]);
+           SWAP(Rot[2][1], Rot[1][2]);
+           maxwell_sym_matrix_rotate(&eps_mean, &eps_mean, Rot);
+#undef SWAP
+
+           /* invert eps_mean to get the Kottke averaged inverse permittivity */
+            maxwell_sym_matrix_invert(md->eps_inv + xyz_index, &eps_mean);
+
 	  }
 	  else { /* undetermined normal vector and/or constant eps */
-	       md->eps_inv[xyz_index] = eps_mean_inv;
+        if (is_interface) { /* if normal is nearly zero but an interface was detected
+                               fall back to ordinary averaging (e.g. for smooth epsilon) */
+           eps_inv_mean = average_eps_inv_over_mesh(r, epsilon, mesh_size, mesh_center,
+                                       m1, m2, m3, mesh_prod_inv, epsilon_data);
```

Comment:
For smooth ε, we don't really need to do averaging at all; the fallback could just evaluate ε at a single point, no?

Reply:
Yeah, I suppose the smooth ε example is particularly apt here (I guess if it is *that* smooth, it will anyway go directly to the next `else` clause due to checking with `SMALL`...).
Something like a symmetric double interface could fit this though (e.g. "ε1 |  ε2 |  ε1" in a single voxel), in principle at least (since it would have zero norm). I guess a situation like that will have other more issues though, so it may not be worth guarding against. What do you prefer?

Reply:
We can leave it as-is, since I doubt you will hit this case very often anyway.

(Even for smoothly varying ε it typically will have a nonzero normal vector.)